File: Managed Source/Arsenal.ImageMounter.Devio/Server/SpecializedProviders/CombinedSeekStream.vb
''''' CombinedSeekStream.vb
''''' 
''''' Copyright (c) 2012-2021, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <http://www.ArsenalRecon.com>
''''' This source code and API are available under the terms of the Affero General Public
''''' License v3.
'''''
''''' Please see LICENSE.txt for full license terms, including the availability of
''''' proprietary exceptions.
''''' Questions, comments, or requests for clarification: http://ArsenalRecon.com/contact/
'''''

Imports System.IO
Imports System.Threading.Tasks
Imports Arsenal.ImageMounter.IO

Namespace Server.SpecializedProviders

    Public Class CombinedSeekStream
        Inherits Stream

        Private ReadOnly _streams As DisposableDictionary(Of Long, Stream)

        Private _current As KeyValuePair(Of Long, Stream)

        Public ReadOnly Property Extendable() As Boolean

        Public ReadOnly Property BaseStreams() As ICollection(Of Stream)
            Get
                Return _streams.Values
            End Get
        End Property

        Public ReadOnly Property CurrentBaseStream() As Stream
            Get
                Return _current.Value
            End Get
        End Property

        Public Sub New()
            Me.New(True)
        End Sub

        Public Sub New(ParamArray ByVal inputStreams() As Stream)
            Me.New(False, inputStreams)
        End Sub

        Public Sub New(ByVal writable As Boolean, ParamArray ByVal inputStreams() As Stream)
            If inputStreams Is Nothing OrElse inputStreams.Length = 0 Then
                _streams = New DisposableDictionary(Of Long, Stream)()

                _Extendable = True
            Else
                _streams = New DisposableDictionary(Of Long, Stream)(inputStreams.Length)

                Array.ForEach(inputStreams, AddressOf AddStream)

                Seek(0, SeekOrigin.Begin)
            End If

            CanWrite = writable
        End Sub

        Private Sub AddStream(ByVal stream As Stream)
            If Not stream.CanSeek OrElse Not stream.CanRead Then
                Throw New NotSupportedException("Needs seekable and readable streams")
            End If

            'INSTANT VB TODO TASK: There is no equivalent to a 'checked' block in VB:
            '			checked
            _Length += stream.Length
            'INSTANT VB TODO TASK: End of the original C# 'checked' block.

            _streams.Add(_Length, stream)
        End Sub

        Public Overrides Function Read(ByVal buffer() As Byte, ByVal index As Integer, ByVal count As Integer) As Integer
            Dim num = 0

            Do While _current.Value IsNot Nothing AndAlso count > 0
                Dim r = _current.Value.Read(buffer, index, count)

                If r <= 0 Then
                    Exit Do
                End If

                Seek(r, SeekOrigin.Current)

                num += r
                index += r
                count -= r
            Loop

            Return num
        End Function

        Public Overrides Sub Write(ByVal buffer() As Byte, ByVal index As Integer, ByVal count As Integer)
            If Not CanWrite Then
                Throw New NotSupportedException()
            End If

            If _position = _Length AndAlso count > 0 AndAlso _Extendable Then
                AddStream(New MemoryStream(buffer, index, count, writable:=True, publiclyVisible:=True))

                Seek(count, SeekOrigin.Current)

                Return
            End If

            If _position >= _Length AndAlso count > 0 Then
                Throw New EndOfStreamException()
            End If

            Do While _current.Value IsNot Nothing AndAlso count > 0
                Dim current_count = CInt(Math.Min(count, _current.Value.Length - _current.Value.Position))

                _current.Value.Write(buffer, index, current_count)

                Seek(current_count, SeekOrigin.Current)

                index += current_count
                count -= current_count
            Loop
        End Sub

        Public Overrides Sub Flush()
            _current.Value?.Flush()
        End Sub

        Private _position As Long

        Public Overrides Property Position() As Long
            Get
                Return _position
            End Get
            Set(ByVal value As Long)
                Seek(value, SeekOrigin.Begin)
            End Set
        End Property

        Public Overrides ReadOnly Property Length() As Long

        Public Overrides Sub SetLength(ByVal value As Long)
            Throw New NotSupportedException()
        End Sub

        Public Overrides ReadOnly Property CanRead() As Boolean
            Get
                Return True
            End Get
        End Property

        Public Overrides ReadOnly Property CanSeek() As Boolean
            Get
                Return True
            End Get
        End Property

        Public Overrides ReadOnly Property CanWrite() As Boolean

        Public Overrides Function Seek(ByVal offset As Long, ByVal origin As SeekOrigin) As Long
            Select Case origin
                Case SeekOrigin.Current
                    offset += _position

                Case SeekOrigin.End
                    offset = Length + offset
            End Select

            If offset < 0 Then
                Throw New ArgumentException("Negative stream positions not supported")
            End If

            _current = _streams.FirstOrDefault(Function(s) s.Key > offset)

            If _current.Value IsNot Nothing Then
                _current.Value.Position = _current.Value.Length - (_current.Key - offset)
            End If

            _position = offset

            Return offset
        End Function

        Public Overrides Sub Close()
            _streams?.Dispose()

            MyBase.Close()
        End Sub
    End Class
End Namespace

File: Managed Source/Arsenal.ImageMounter/IO/Streams/CombinedSeekStream.cs
//  
//  Copyright (c) 2012-2025, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <http://www.ArsenalRecon.com>
//  This source code and API are available under the terms of the Affero General Public
//  License v3.
// 
//  Please see LICENSE.txt for full license terms, including the availability of
//  proprietary exceptions.
//  Questions, comments, or requests for clarification: http://ArsenalRecon.com/contact/
// 

using DiscUtils.Streams.Compatibility;
using LTRData.Extensions.Async;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

#pragma warning disable IDE0079 // Remove unnecessary suppression

namespace Arsenal.ImageMounter.IO.Streams;

public class CombinedSeekStream : CompatibilityStream
{
    private readonly List<KeyValuePair<long, Stream>> streams;

    private KeyValuePair<long, Stream> current;

    public bool Extendable { get; }

    public IReadOnlyCollection<KeyValuePair<long, Stream>> BaseStreams => streams;

    public Stream CurrentBaseStream => current.Value;

    public CombinedSeekStream()
        : this(true)
    {
    }

    public CombinedSeekStream(params Stream[] inputStreams)
        : this(false, inputStreams)
    {
    }

    public CombinedSeekStream(IEnumerable<Stream> inputStreams)
        : this(false, inputStreams)
    {
    }

    public CombinedSeekStream(bool writable, params Stream[] inputStreams)
    {
        if (inputStreams is null || inputStreams.Length == 0)
        {
            streams = [];

            Extendable = true;
        }
        else
        {
            streams = new(inputStreams.Length);

            Array.ForEach(inputStreams, AddStream);

            Seek(0, SeekOrigin.Begin);
        }

        CanWrite = writable;
    }

    public CombinedSeekStream(bool writable, IEnumerable<Stream> inputStreams)
    {
        streams = [];

        foreach (var stream in inputStreams)
        {
            AddStream(stream);
        }

        if (streams.Count == 0)
        {
            Extendable = true;
        }
        else
        {
            Seek(0, SeekOrigin.Begin);
        }

        CanWrite = writable;
    }

    public void AddStream(Stream stream)
    {
        if (!stream.CanSeek || !stream.CanRead)
        {
            throw new NotSupportedException("Needs seekable and readable streams");
        }

        if (stream.Length == 0)
        {
            return;
        }

        checked
        {
            _length += stream.Length;
        }

        streams.Add(new(_length, stream));
    }

    public override int Read(byte[] buffer, int index, int count)
    {
        var num = 0;

        while (current.Value is not null && count > 0)
        {
            var r = current.Value.Read(buffer, index, count);

            if (r <= 0)
            {
                break;
            }

            Seek(r, SeekOrigin.Current);

            num += r;
            index += r;
            count -= r;
        }

        return num;
    }

    public override async Task<int> ReadAsync(byte[] buffer, int index, int count, CancellationToken cancellationToken)
    {
        var num = 0;

        while (current.Value is not null && count > 0)
        {
            var r = await current.Value.ReadAsync(buffer.AsMemory(index, count), cancellationToken).ConfigureAwait(false);

            if (r <= 0)
            {
                break;
            }

            Seek(r, SeekOrigin.Current);

            num += r;
            index += r;
            count -= r;
        }

        return num;
    }

    public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state) =>
        ReadAsync(buffer, offset, count, CancellationToken.None).AsAsyncResult(callback, state);

    public override int EndRead(IAsyncResult asyncResult) =>
        ((Task<int>)asyncResult).GetAwaiter().GetResult();

    public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
    {
        var count = buffer.Length;
        var index = 0;
        var num = 0;

        while (current.Value is not null && count > 0)
        {
            var r = await current.Value.ReadAsync(buffer.Slice(index, count), cancellationToken).ConfigureAwait(false);

            if (r <= 0)
            {
                break;
            }

            Seek(r, SeekOrigin.Current);

            num += r;
            index += r;
            count -= r;
        }

        return num;
    }

    public override int Read(Span<byte> buffer)
    {
        var count = buffer.Length;
        var num = 0;

        while (current.Value is not null && count > 0)
        {
            var r = current.Value.Read(buffer.Slice(num, count));

            if (r <= 0)
            {
                break;
            }

            Seek(r, SeekOrigin.Current);

            num += r;
            count -= r;
        }

        return num;
    }

    public override void Write(byte[] buffer, int index, int count)
    {
        if (!CanWrite)
        {
            throw new NotSupportedException();
        }

        if (position == _length && count > 0 && Extendable)
        {
            AddStream(new MemoryStream(buffer, index, count, writable: true, publiclyVisible: true));

            Seek(count, SeekOrigin.Current);

            return;
        }

        if (position >= _length && count > 0)
        {
            throw new EndOfStreamException();
        }

        while (current.Value is not null && count > 0)
        {
            var current_count = (int)Math.Min(count, current.Value.Length - current.Value.Position);

            current.Value.Write(buffer, index, current_count);

            Seek(current_count, SeekOrigin.Current);

            index += current_count;
            count -= current_count;
        }
    }

    public override async Task WriteAsync(byte[] buffer, int index, int count, CancellationToken cancellationToken)
    {
        if (!CanWrite)
        {
            throw new NotSupportedException();
        }

        if (position == _length && count > 0 && Extendable)
        {
            AddStream(new MemoryStream(buffer, index, count, writable: true, publiclyVisible: true));

            Seek(count, SeekOrigin.Current);

            return;
        }

        if (position >= _length && count > 0)
        {
            throw new EndOfStreamException();
        }

        while (current.Value is not null && count > 0)
        {
            var current_count = (int)Math.Min(count, current.Value.Length - current.Value.Position);

            await current.Value.WriteAsync(buffer.AsMemory(index, current_count), cancellationToken).ConfigureAwait(false);

            Seek(current_count, SeekOrigin.Current);

            index += current_count;
            count -= current_count;
        }
    }

    public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state) =>
        WriteAsync(buffer, offset, count, CancellationToken.None).AsAsyncResult(callback, state);

    public override void EndWrite(IAsyncResult asyncResult) =>
        ((Task)asyncResult).GetAwaiter().GetResult();

    public override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
    {
        if (!CanWrite)
        {
            throw new NotSupportedException();
        }

        var count = buffer.Length;
        var index = 0;

        if (position == _length && count > 0 && Extendable)
        {
            AddStream(new MemoryStream(buffer.ToArray(), index, count, writable: true, publiclyVisible: true));

            Seek(count, SeekOrigin.Current);

            return;
        }

        if (position >= _length && count > 0)
        {
            throw new EndOfStreamException();
        }

        while (current.Value is not null && count > 0)
        {
            var current_count = (int)Math.Min(count, current.Value.Length - current.Value.Position);

            await current.Value.WriteAsync(buffer.Slice(index, current_count), cancellationToken).ConfigureAwait(false);

            Seek(current_count, SeekOrigin.Current);

            index += current_count;
            count -= current_count;
        }
    }

    public override void Write(ReadOnlySpan<byte> buffer)
    {
        if (!CanWrite)
        {
            throw new NotSupportedException();
        }

        var count = buffer.Length;
        var index = 0;

        if (position == _length && count > 0 && Extendable)
        {
            AddStream(new MemoryStream(buffer.ToArray(), index, count, writable: true, publiclyVisible: true));

            Seek(count, SeekOrigin.Current);

            return;
        }

        if (position >= _length && count > 0)
        {
            throw new EndOfStreamException();
        }

        while (current.Value is not null && count > 0)
        {
            var current_count = (int)Math.Min(count, current.Value.Length - current.Value.Position);

            current.Value.Write(buffer.Slice(index, current_count));

            Seek(current_count, SeekOrigin.Current);

            index += current_count;
            count -= current_count;
        }
    }

    public override void Flush() => current.Value?.Flush();

    public override Task FlushAsync(CancellationToken cancellationToken)
        => current.Value?.FlushAsync(cancellationToken) ?? Task.CompletedTask;

    private long position;

    public override long Position
    {
        get => position;
        set => Seek(value, SeekOrigin.Begin);
    }

    public override long? GetPositionInBaseStream(Stream baseStream, long virtualPosition)
    {
        if (ReferenceEquals(baseStream, this))
        {
            return virtualPosition;
        }

        var stream = streams.FirstOrDefault(s => s.Key > virtualPosition);

        if (stream.Value is null)
        {
            throw new EndOfStreamException();
        }

        var subStreamPosition = virtualPosition - (stream.Key - stream.Value.Length);

        return stream.Value.GetPositionInBaseStream(baseStream, subStreamPosition);
    }

    private long _length;

    public override long Length => _length;

    public override void SetLength(long value) => throw new NotSupportedException();

    public override bool CanRead => true;

    public override bool CanSeek => true;

    public override bool CanWrite { get; }

    public override long Seek(long offset, SeekOrigin origin)
    {
        switch (origin)
        {
            case SeekOrigin.Current:
                offset += position;
                break;

            case SeekOrigin.End:
                offset += Length;
                break;
        }

        if (offset < 0)
        {
            throw new ArgumentException("Negative stream positions not supported");
        }

        current = streams.FirstOrDefault(s => s.Key > offset);

        current.Value?.Position = current.Value.Length - (current.Key - offset);

        position = offset;

        return offset;
    }

    protected override void Dispose(bool disposing)
    {
        if (streams != null)
        {
            if (disposing)
            {
                streams.AsParallel().ForAll(stream => stream.Value.Dispose());
            }

            streams.Clear();
        }

        base.Dispose(disposing);
    }
}
